### ISSUE TITLE
Feature: LED widgets

### ISSUE THREAD

**Issue by crides** [zmkfirmware/zmk#1003]
This is sort of a term I coined, but the idea is to have LEDs act as displays. The widgets would be parameterized by PWM brightness, LED number (as in which LED to use; like colors or just linear position), and on/off length. The basic features would just be an API to preset several widgets and display them (with some spacing between them) in order when they are requested.

Just like actual screen display widgets, this'll be hard to customize for various different events so we shouldn't go there (unless someone wants to group all core events into a 32bit flag, which would be cool but unnecessary).

This is to mainly put the idea out here. I will need this at some point in the near future (need as in nice to have), so if nobody else want this I'll probably implement this anyway (just really not sure _when_ it will be); otherwise, others can take a stab if they want to

**Comment by idesignstuff** [comment]
I have Xiao BLE (NRF) boards with an integrated common cathode RGB light. I'd really like to be able to change the PWM values for each color based on behaviors like 

- when I change layers
- when caps lock is on
- when the battery is low
- when the bt is connected, etc.
- 
This looks like it may work for that too.

**Comment by crides** [comment]
yes, you are probably looking for the PR itself, which is linked

**Comment by sporkus** [comment]
> yes, you are probably looking for the PR itself, which is linked

Hi, thanks for your work on led widgets. I was really hoping to get some simple indicator working with the xiao BLE. From the datasheet, it looks like none of the 3-in-1 led can be controlled by the pwm. As seen here: https://files.seeedstudio.com/wiki/XIAO-BLE/XIAO-nRF52840-pinout_sheet.xlsx, all three channels are simple GPIO only.

I tried fllowoioi

**Comment by crides** [comment]
You should be able to get PWM on all GPIOs. Having your config/code be available somewhere would help

**Comment by sporkus** [comment]
> You should be able to get PWM on all GPIOs. Having your config/code be available somewhere would help

Thanks for the comment. Looking at the source code, it looks like it's just calling [`led_set_brightness()`](https://docs.zephyrproject.org/latest/hardware/peripherals/led.html#group__led__interface_1gaca479fd77518331f4fc84f788e345882) on the LED. It also says it should turn the LED on if value/brightness is non-zero, so your existing code/config "should" also work for non-PWM LEDs without any change.

I set up a local zephyr build environment and was able to build the blinky / blinky PWM sample. For the basic blinky, I was able to get all three colors working. But the PWM example seem to only affect the very dim charge LED. I went through the zephyr doc specific to the [nrf](https://docs.zephyrproject.org/latest/build/dts/api/bindings/pwm/nordic%2Cnrf-pwm.html#dtbinding-nordic-nrf-pwm) and couldn't really make sense of how to set up the pwm for the other pin. As the 3-in-1 led isn't very bright, I'd like to just get them toggle on/off for now.

I have a minimal branch here for testing ([compare view](https://github.com/sporkus/zmk-config/compare/led-widget)), and I am getting this build error:
```
 /__w/zmk-config/zmk-config/zmk/app/src/led_widgets.c:66: undefined reference to `led_widgets'
/opt/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /__w/zmk-config/zmk-config/zmk/app/src/led_widgets.c:66: undefined reference to `__device_dts_ord_101'
```
Here's the list of things I'd done:

- merge your branch into my zmk fork
- point the build to the new led-widget branch
- configured the shield with `CONFIG_ZMK_LED_WIDGETS`
- added a user defined `led_widgets.c` with just the boot event for the shield, and added a CMakeLists file
- added 3 GPIO LEDs and the `led-widgets` node to the devicetree 

As far as I can tell, I have everything configured correctly. Clearly, I must be missing something, and I'd greatly appreciate any advice.

**Comment by crides** [comment]
So the device didn't actually get enabled. Check `menuconfig` and see why that is maybe?

**Comment by crides** [comment]
oh it's a shield. I don't think C files are supported in shields (in zmk at least)

**Comment by sporkus** [comment]
> oh it's a shield. I don't think C files are supported in shields (in zmk at least)

It seems like it. It compiled even when I added with some bogus code in c file and added non-existing sources to the shield. I'll see how to port the config off to dts.

**Comment by crides** [comment]
I didn't figure out how to fetch the config from the dts. The macro shenenigans just gets too crazy. The dts chunk of led widgets config there is just for testing. it doesn't do anything at this point
